File: Data Structures & Algorithms/binary-tree-diameter/submission-0.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number}
     */
    height(root){
        if(root==null) return 0;
        this.height(root.left);
        this.height(root.right);
        return Math.max(this.height(root.left),this.height(root.right)) +1;
    }
    diameterOfBinaryTree(root) {
        if(root==null) return 0;
        let maxheightLeft= this.height(root.left);
        let maxHeightRight = this.height(root.right);
        
        let leftDiameter = this.diameterOfBinaryTree(root.left);
        let rightDiameter = this.diameterOfBinaryTree(root.right);

        return Math.max(maxheightLeft+maxHeightRight, leftDiameter, rightDiameter );

    }
}

File: Data Structures & Algorithms/trapping-rain-water/submission-0.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int water=0;
        int left = 0;
        int right = height.size()-1;
        int leftMax= 0;
        int rightMax = 0;
        while(left<=right){
            if(height[left]<height[right]){
                if(height[left]>leftMax){
                    leftMax= max(leftMax, height[left]);
                }else{
                    water += leftMax - height[left];
                }
                left++;
            }else if(height[left]>=height[right]){
                if(height[right]>rightMax){
                    rightMax= max(rightMax, height[right]);
                }else{
                    water += rightMax - height[right];
                }
                right--;
            }
        }
        return water;
    }
};
